src/jit/importer.cpp

```diff
@@ -16440,6 +16439,33 @@ bool Compiler::impReturnInstruction(BasicBlock* block, int prefixFlags, OPCODE&
 #endif
                 }
 
+                // If we are inlining a method that returns a struct byref, check whether we are "reinterpreting" the
+                // struct.
+                if ((returnType == TYP_BYREF) && (info.compRetType == TYP_BYREF) && (op2->OperGet() == GT_ADDR))
+                {
+                    GenTree* op2Child = op2->gtGetOp1();
+                    if ((op2Child->OperGet() == GT_LCL_VAR) && varTypeIsStruct(op2Child->TypeGet()))
+                    {
+                        LclVarDsc* varDsc = &lvaTable[op2Child->gtLclVarCommon.gtLclNum];
+
+                        CORINFO_CLASS_HANDLE referentClassHandle;
+                        CorInfoType          referentType =
+                            info.compCompHnd->getChildType(info.compMethodInfo->args.retTypeClass,
+                                                           &referentClassHandle);
+                        if (varTypeIsStruct(JITtype2varType(referentType)) &&
+                            (varDsc->lvVerTypeInfo.GetClassHandle() != referentClassHandle))
+                        {
+                            // We are returning a byref to struct1; the method signature specifies return type as byref
+                            // to struct2. struct1 and struct2 are different so we are "reinterpreting" the struct.
+                            // This may happen in, for example, System.Runtime.CompilerServices.Unsafe.As<TFrom, TTo>.
+                            // We need to mark the source struct variable as having overlapping fields because its
+                            // fields may be accessed using field handles of a different type, which may confuse
+                            // optimizations, in particular, value numbering.
+                            varDsc->lvOverlappingFields = true;
```

Comment:
You should add a JITDUMP message here indicating that you have set lvOverlappingFields

Reply:
Done.